DOCS-11871: Provided some more detail on batch tuning (#2023)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/tuning-batch-processing.adoc

Additionally, batch processing requires having enough memory available to process the threads in parallel, which means moving the records from persistent storage into RAM in a fixed-size block. The larger your records and their quantity, the more available memory you need for batch processing.

> == Batch Block size

Reviewer: ```suggestion
== Batch Block Size
```

> == Batch Block Size

Reviewer: ```suggestion
== Batch Block Size

```

<batch:job jobName="test-batch" blockSize="${batch.block.size}">
----

> == Max Concurrency

Reviewer: ```suggestion
== Max Concurrency

```

The maximum number of threads that can be used to process a batch scope is by default set to 2 times the cores the JVM detects (note that on fractional core deployments such as CloudHub and Runtime Fabric this figure will not be a fractional amount, rather what the JVM detects). This can be manually set to higher or lower than the automatic figure. It is possible that for a given batch job instance there is insufficient records to warrant this full amount, so this is an upper bound on the number of threads that may be used.

The follow example shows the maxConcurrency overridden from the default automatic size to one loaded from a property:

[source,xml,linenums]
----
<batch:job jobName="an-example-batch" maxConcurrency="${batch.block.maxConcurrency}">
----

> == Batch Aggregators

Reviewer: ```suggestion
== Batch Aggregators

```

<batch:job jobName="test-batch" blockSize="${batch.block.size}">
----

== Max Concurrency

> The maximum number of threads that can be used to process a batch scope is by default set to 2 times the cores the JVM detects (note that on fractional core deployments such as CloudHub and Runtime Fabric this figure will not be a fractional amount, rather what the JVM detects). This can be manually set to higher or lower than the automatic figure. It is possible that for a given batch job instance there is insufficient records to warrant this full amount, so this is an upper bound on the number of threads that may be used.

Reviewer: ```suggestion
The maximum number of threads for processing a Batch scope defaults to 2 times the number of cores that the JVM detects. On fractional core deployments, such as CloudHub and Anypoint Runtime Fabric, the JVM detects a _non-fractional_ number of cores. You can manually increase or lower the automated concurrency setting. The Max Concurrency setting enables you to set an upper bound on the number of threads in case a given batch job instance has insufficient records to warrant the use of more threads.
```

<batch:job jobName="test-batch" blockSize="${batch.block.size}">
----

== Max Concurrency

> The maximum number of threads for processing a Batch scope defaults to 2 times the number of cores that the JVM detects. On fractional core deployments, such as CloudHub and Anypoint Runtime Fabric, the JVM detects a _non-fractional_ number of cores. You can manually increase or lower the automated concurrency setting. The Max Concurrency setting enables you to set an upper bound on the number of threads in case a given batch job instance has insufficient records to warrant the use of more threads.

Reviewer: ```suggestion
The maximum number of threads for processing a Batch scope defaults to 2 times the number of cores that the JVM detects. On fractional core deployments, such as CloudHub and Anypoint Runtime Fabric, the JVM detects a _non-fractional_ number of cores. You can manually increase or lower the automated concurrency setting. The Max Concurrency setting enables you to set an upper limit on the number of threads to use for a batch job instance that, for example, has insufficient records to warrant more threads.
```

> The follow example shows the maxConcurrency overridden from the default automatic size to one loaded from a property:

Reviewer: ```suggestion
The follow example overrides the automated `maxConcurrency` default to a setting that is loaded from a property:
```

[source,xml,linenums]
----
<batch:job jobName="an-example-batch" maxConcurrency="${batch.block.maxConcurrency}">
----

== Batch Aggregators

> One million round trips across a network impacts performance, whether on single or multiple threads. Communication with SaaS systems can easily consume the quota of API calls. Instead of processing records one by one, you can improve performance by using the Batch Aggregator scope to perform bulk operations on groups of records. The scope uses an Aggregator Size setting to divide the records into groups. 

Reviewer: ```suggestion
One million round trips across a network impacts performance, whether on single or multiple threads. Communication with SaaS systems can easily consume the quota of API calls. Instead of processing records one by one, you can improve performance by using the Batch Aggregator scope to perform bulk operations on arrays of records. The scope uses an Aggregator Size setting to limit the number of records in each array. 
```

> Another factor impacting performance of batch processes is whether you want to do operations on sets of records rather than one by one. For instance: 1 million round trips across a network, even using multiple threads, is still 1 million round trips. When talking to SaaS systems this might easily consume the quota of API calls. To improve the "network chattiness" a batch aggregator with a fixed size will mean bulk operations can be used. The number of records will then be a parameter which can be tuned, mindful of the maximum allowed by whatever operations you are performing (for example: you may only be able to do up to X records at a time for a given system, so should try values below this during performance testing). Note that logic should handle any number of records up to and including the specified size as the overall record number may not divide neatly into that amount and need to process the remainder.

Reviewer: ```suggestion
One million round trips across a network, whether on single or multiple threads, impacts performance. Communication with SaaS systems can easily consume the quota of API calls. Instead of processing records one by one, you can improve performance by using the Batch Aggregator scope to perform bulk operations on groups of records. The scope uses an Aggregator Size setting to divide the records into groups. When determining the number of records to process in bulk, you must take into account the maximum number of records permitted by operations on the system with which the communication takes place. It is important to conduct performance testing on your Aggregator Size setting _and on lower values_. Your logic must handle any number of records up to and including the value of your setting because the number of records that the system receives can be lower than your specified value. For example, if there are 120 records, and your Aggregator Size setting is 50, the last submission is 20 records.
```

> When determining the number of records to process in bulk, you must take into account the maximum number of records permitted by operations on the system with which the communication takes place. It is important to conduct performance testing on your Aggregator Size setting _and on lower values_. Your logic must handle any number of records up to and including the value of your setting because the number of records that the system receives can be lower than your specified value. For example, if there are 120 records, and your Aggregator Size setting is 50, the last submission is 20 records.

Reviewer: ```suggestion
When determining the number of records to process in an array of records, you must take into account the maximum number of records permitted by operations on the system with which the communication takes place. It is important to conduct performance testing on your Aggregator Size setting _and on lower values_. Your logic must handle any number of records up to and including the value of your setting because the number of records that the system receives can be lower than your specified value. For example, if there are 120 records, and your Aggregator Size setting is 50, the last submission is 20 records.
```

Author: Hrm.. Not really common term to have the term "group".. We usually talk about bulk operations or bulk processing. Maybe a "chunk of records" "A collection" or "an array of records" or something. "Group" as a term only really gets used for "group by" which is a bit of a different thing.

> When determining the number of records to process in a group, you must take into account the maximum number of records permitted by operations on the system with which the communication takes place. It is important to conduct performance testing on your Aggregator Size setting _and on lower values_. Your logic must handle any number of records up to and including the value of your setting because the number of records that the system receives can be lower than your specified value. For example, if there are 120 records, and your Aggregator Size setting is 50, the last submission is 20 records.

Reviewer: ```suggestion
When determining the number of records to process in an array of records, you must take into account the maximum number of records permitted by operations on the system with which the communication takes place. It is important to conduct performance testing on your Aggregator Size setting _and on lower values_. Your logic must handle any number of records up to and including the value of your setting because the number of records that the system receives can be lower than your specified value. For example, if there are 120 records, and your Aggregator Size setting is 50, the last submission is 20 records.
```

> The following example shows a batch step with a batch step aggregator set to a fixed size of 50:

Reviewer: ```suggestion
The following example shows a Batch Step scope with a Batch Aggregator scope that is set to a fixed size of `50`:

```

[source,xml,linenums]
----
<batch:step name="Batch_Step_PushToXYZ" >
  <batch:aggregator doc:name="Batch Aggregator" size="50">
     <!-- code for an array of 50 (or fewer) records -->
  </batch:aggregator>
</batch:step>
----

> If the aggregator contains an operation that streams data, you can activate the scope's Streaming option  (`streaming="true"`) instead of using an Aggregator Size setting (`size="50"`).

Reviewer: ```suggestion
If the aggregator contains an operation that streams data, you can activate the scope's Streaming option (`streaming="true"`) instead of using an Aggregator Size setting (`size="50"`).
```

If the aggregator contains an operation that streams data, you can activate the scope's Streaming option (`streaming="true"`) instead of using an Aggregator Size setting (`size="50"`).

> == Multiple Concurrent Instances of a Batch Job Scope

Reviewer: ```suggestion
== Multiple Concurrent Instances of a Batch Job
```

> Many use cases for Batch Job scopes process only one request (batch job instance) at a time, but some cases simultaneously execute two or more batch job instances. To tune the thread pool sharing in such cases, you can set a scheduling strategy: 

Reviewer: ```suggestion
Many use cases for Batch Job scopes process only one request (batch job instance) at a time, but some cases simultaneously execute two or more batch job instances. To tune the thread pool sharing in such cases, you can set a scheduling strategy, each of which is tracked separately by its batch instance ID: 
```

> Many use cases for Batch Job scopes process only one request (batch job instance) at a time, but some cases simultaneously execute two or more batch job instances. To tune the thread pool sharing in such cases, you can set a scheduling strategy, each of which is tracked separately by its batch instance ID: 

Reviewer: ```suggestion
Many use cases for Batch Job scopes process only one request (batch job instance) at a time, but some cases simultaneously execute two or more batch job instances. To tune the thread pool sharing in such cases, you can set one of the following scheduling strategies, each of which is tracked separately by its batch instance ID: 
```

The following example shows a Batch Step scope with a Batch Aggregator scope that is set to a fixed size of `50`:

[source,xml,linenums]
----
<batch:step name="Batch_Step_PushToXYZ" >
  <batch:aggregator doc:name="Batch Aggregator" size="50">

>      <!-- code to work with array of 50 (or also smaller) records -->

Reviewer: ```suggestion
     <!-- code for an array of 50 (or fewer) records -->
```

  </batch:aggregator>
</batch:step>
----

> If the aggregator contains an operation that streams data, you can set the activate the scope's Streaming option instead of using an Aggregator Size setting.

Reviewer: ```suggestion
If the aggregator contains an operation that streams data, you can set the activate the scope's Streaming option  (`streaming="true"`) instead of using an Aggregator Size setting (`size="50"`).
```

> If the aggregator contains an operation that streams data, you can set the activate the scope's Streaming option, which is independent of the Aggregator Size setting.

Reviewer: ```suggestion
If the aggregator contains an operation that streams data, you can set the activate the scope's Streaming option instead of using an Aggregator Size setting.
```

> If there is a streaming operation possible for the records at that point the "streaming" mode for aggregator can be used. This is mutually exclusive to the fixed size.

Reviewer: ```suggestion
If the aggregator contains an operation that streams data, you can set the activate the scope's Streaming option, which is independent of the Aggregator Size setting.
```

> If the aggregator contains an operation that streams data, you can set the activate the scope's Streaming option  (`streaming="true"`) instead of using an Aggregator Size setting (`size="50"`).

Reviewer: ```suggestion
If the aggregator contains an operation that streams data, you can activate the scope's Streaming option  (`streaming="true"`) instead of using an Aggregator Size setting (`size="50"`).
```

> == Multiple concurrent instances of a Batch job scope

Reviewer: ```suggestion
== Multiple Concurrent Instances of a Batch Job Scope
```

> Many use cases for Batch Job scopes process only one request (batch job instance) at a time, but some cases simultaneously execute two or more batch job instances. To tune the thread pool sharing in such cases, you can set a scheduling strategy. The default is `ORDERED_SEQUENTIAL`. `ROUND_ROBIN` is available as an alternative.

Reviewer: ```suggestion
Many use cases for Batch Job scopes process only one request (batch job instance) at a time, but some cases simultaneously execute two or more batch job instances. To tune the thread pool sharing in such cases, you can set a scheduling strategy: 
```

> Many use cases for Batch Job scopes process only one request (batch job instance) at a time, but some cases simultaneously execute two or more batch job instances. To tune the thread pool sharing in such cases, you can set a scheduling strategy. The default is `ORDERED_SEQUENTIAL`.  `ROUND_ROBIN` is available as an alternative.

Reviewer: ```suggestion
Many use cases for Batch Job scopes process only one request (batch job instance) at a time, but some cases simultaneously execute two or more batch job instances. To tune the thread pool sharing in such cases, you can set a scheduling strategy. The default is `ORDERED_SEQUENTIAL`. `ROUND_ROBIN` is available as an alternative.
```

> Many use cases for batch job scopes are only ever going to have one request (batch job instance) at a time to process, but some might involve two or more batch job instances executed simultaneously. To allow for tuning of the behaviour of the sharing of the thread pool in this situation the "scheduling strategy" can be set. The default is "ORDERED_SEQUENTIAL" but a "ROUND_ROBIN" is available as an alternative.

Reviewer: ```suggestion
Many use cases for Batch Job scopes process only one request (batch job instance) at a time, but some cases simultaneously execute two or more batch job instances. To tune the thread pool sharing in such cases, you can set a scheduling strategy. The default is `ORDERED_SEQUENTIAL`.  `ROUND_ROBIN` is available as an alternative.
```

> Ordered sequential (the default) is going to limit the impact of concurrent instances by processing them in the order they were triggered. This is useful when non-sequential processing of the same batch job might cause issues such as data consistency or hit resource limitations. 

Reviewer: ```suggestion
`ORDERED_SEQUENTIAL`: Default scheduling strategy that limits the impact of concurrent instances by processing the instances in the order they are triggered. This strategy is useful when non-sequential processing of the same batch job can cause issues, such as data inconsistency, or reach resource limitations. 
```

> Round robin will attempt to share the threads amongst whatever batch job instances are "in flight". 

Reviewer: ```suggestion
`ROUND_ROBIN`: Scheduling strategy that shares the threads with batch job instances that are "in flight". 
```

> Both options for scheduling strategy (Round robin or Ordered sequential) will continue to be tracked separately via their batch instance ID. 

Reviewer: ```suggestion
```
